unified_planning/model/action.py

```diff
@@ -764,3 +764,228 @@ def __repr__(self) -> str:
+        assert (
+            effect.environment == self._environment
+        ), "effect does not have the same environment of the action"
+
+        self._effects.append(effect)
+
+    def add_precondition(
+        self,
+        precondition: Union[
+            "up.model.fnode.FNode",
+            "up.model.fluent.Fluent",
+            "up.model.parameter.Parameter",
+            bool,
+        ],
+    ):
+        """
+        Adds the given expression to `action's preconditions`.
+
+        :param precondition: The expression that must be added to the `action's preconditions`.
+        """
+        (precondition_exp,) = self._environment.expression_manager.auto_promote(
+            precondition
+        )
+        assert self._environment.type_checker.get_type(precondition_exp).is_bool_type()
+        if precondition_exp == self._environment.expression_manager.TRUE():
+            return
+        free_vars = self._environment.free_vars_oracle.get_free_variables(
+            precondition_exp
+        )
+        if len(free_vars) != 0:
+            raise UPUnboundedVariablesError(
+                f"The precondition {str(precondition_exp)} has unbounded variables:\n{str(free_vars)}"
+            )
+        if precondition_exp not in self._preconditions:
+            self._preconditions.append(precondition_exp)
+
+    def add_derivative(
+        self,
+        fluent: Union["up.model.fnode.FNode", "up.model.fluent.Fluent"],
+        value: "up.model.expression.Expression",
+    ):
+        """
+        Adds the given `derivative effect` to the `process's effects`.
+
+        :param fluent: The `fluent` objective of the time derivative definition.
+        :param value: The given `fluent` is incremented by the given `value`.
+        """
+        (
+            fluent_exp,
+            value_exp,
+            condition_exp,
+        ) = self._environment.expression_manager.auto_promote(
+            fluent,
+            value,
+            True,
+        )
+        if not fluent_exp.is_fluent_exp() and not fluent_exp.is_dot():
+            raise UPUsageError(
+                "fluent field of add_increase_effect must be a Fluent or a FluentExp or a Dot."
+            )
+        if not fluent_exp.type.is_compatible(value_exp.type):
+            raise UPTypeError(
+                f"Process effect has an incompatible value type. Fluent type: {fluent_exp.type} // Value type: {value_exp.type}"
+            )
+        if not fluent_exp.type.is_int_type() and not fluent_exp.type.is_real_type():
+            raise UPTypeError("Derivative can be created only on numeric types!")
+        self._add_effect_instance(
+            up.model.effect.Effect(
+                fluent_exp,
+                value_exp,
+                condition_exp,
+                kind=up.model.effect.EffectKind.INCREASE,
+                forall=tuple(),
+            )
+        )
+
+
+class Event(InstantaneousAction):
```

Comment:
This is kind of dangerous: beware that `isinstance(x, InstantaneousAction)` would return `True` if x is an Event. I get that we save some code duplication, but here I would inherit from `Action` only... Then, maybe we can create a mixin to avoid the code duplication

Reply:
For example, https://github.com/aiplan4eu/unified-planning/blob/master/unified_planning/model/mixins/actions_set.py#L60 is probably broken by this choice

Reply:
yep